<p style="text-align:center">
    <a href="https://skills.network" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="300" alt="Skills Network Logo">
    </a>
</p>


# Practice Project: GDP Data extraction and processing

Estimated time needed: **30** minutes

## Introduction

In this practice project, you will put the skills acquired through the course to use. You will extract data from a website using webscraping and reqeust APIs process it using Pandas and Numpy libraries.


## Project Scenario:

An international firm that is looking to expand its business in different countries across the world has recruited you. You have been hired as a junior Data Engineer and are tasked with creating a script that can extract the list of the top 10 largest economies of the world in descending order of their GDPs in Billion USD (rounded to 2 decimal places), as logged by the International Monetary Fund (IMF). 

The required data seems to be available on the URL mentioned below:


URL: https://web.archive.org/web/20230902185326/https://en.wikipedia.org/wiki/List_of_countries_by_GDP_%28nominal%29


## Objectives

After completing this lab you will be able to:

 - Use Webscraping to extract required information from a website.
 - Use Pandas to load and process the tabular data as a dataframe.
 - Use Numpy to manipulate the information contatined in the dataframe.
 - Load the updated dataframe to CSV file.


---


## Dislcaimer

If you are using a downloaded version of this notebook on your local machine, you may encounter a warning message as shown in the screenshot below.

<p style="text-align:center">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/mod_5/practice_project_disclaimer.png" width="700" alt="warning message">
</p>


This does not affect the execution of your codes in any way and can be simply ignored. 


# Setup


For this lab, we will be using the following libraries:

*   [`pandas`](https://pandas.pydata.org/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMML0187ENSkillsNetwork31430127-2021-01-01) for managing the data.
*   [`numpy`](https://numpy.org/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMML0187ENSkillsNetwork31430127-2021-01-01) for mathematical operations.


### Importing Required Libraries

_We recommend you import all required libraries in one place (here):_


In [1]:
import numpy as np
import pandas as pd

# You can also use this section to suppress warnings generated by your code:
def warn(*args, **kwargs):
    pass
import warnings
warnings.warn = warn
warnings.filterwarnings('ignore')

---


# Exercises

### Exercise 1
Extract the required GDP data from the given URL using Web Scraping.


In [2]:
URL="https://web.archive.org/web/20230902185326/https://en.wikipedia.org/wiki/List_of_countries_by_GDP_%28nominal%29"

You can use Pandas library to extract the required table directly as a DataFrame. Note that the required table is the third one on the website, as shown in the image below.

<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/images/pandas_wbs_3.png">


In [3]:
#Importing the necessary libraries
import numpy as np
import pandas as pd

#Extract tables from webpage using Pandas
tables = pd.read_html('https://en.wikipedia.org/wiki/List_of_countries_by_GDP_(nominal)')
df = tables[2]
df.columns = range(df.shape[1])
df = df[[0,2]]
df = df.iloc[1:21,:]
df.columns = ['Country','GDP (Million USD)']
df

Country GDP (Million USD)
1    United States          28781083
2            China          18532633
3          Germany           4591100
4            Japan           4110452
5            India           3937011
6   United Kingdom           3495261
7           France           3130014
8           Brazil           2331391
9            Italy           2328028
10          Canada           2242182
11          Russia           2056844
12          Mexico           2017025
13       Australia           1790348
14     South Korea           1760947
15           Spain           1647114
16       Indonesia           1475690
17     Netherlands           1142513
18          Turkey           1113561
19    Saudi Arabia           1106015
20     Switzerland            938458

### Exercise 2
Modify the GDP column of the DataFrame, converting the value available in Million USD to Billion USD. Use the `round()` method of Numpy library to round the value to 2 decimal places. Modify the header of the DataFrame to `GDP (Billion USD)`.


In [4]:
# Importing the necessary libraries
import numpy as np
import pandas as pd

# Extract tables from webpage using Pandas
tables = pd.read_html('https://en.wikipedia.org/wiki/List_of_countries_by_GDP_(nominal)')
df = tables[2]
df.columns = range(df.shape[1])
df = df[[0, 2]]
df = df.iloc[1:21, :]
df.columns = ['Country', 'GDP (Million USD)']

# Rename the column
df.rename(columns={'GDP (Million USD)': 'GDP (Billion USD)'}, inplace=True)

# Convert 'GDP (Billion USD)' column to float
df['GDP (Billion USD)'] = df['GDP (Billion USD)'].astype(float)

# Convert to billion USD and round to 2 decimal places
df['GDP (Billion USD)'] /= 1000
df['GDP (Billion USD)'] = df['GDP (Billion USD)'].round(2)

#Setting display format with commas for easier reading
pd.options.display.float_format = '{:,.2f}'.format
df


Country  GDP (Billion USD)
1    United States          28,781.08
2            China          18,532.63
3          Germany           4,591.10
4            Japan           4,110.45
5            India           3,937.01
6   United Kingdom           3,495.26
7           France           3,130.01
8           Brazil           2,331.39
9            Italy           2,328.03
10          Canada           2,242.18
11          Russia           2,056.84
12          Mexico           2,017.02
13       Australia           1,790.35
14     South Korea           1,760.95
15           Spain           1,647.11
16       Indonesia           1,475.69
17     Netherlands           1,142.51
18          Turkey           1,113.56
19    Saudi Arabia           1,106.02
20     Switzerland             938.46

#### There is almost complete, but we have a problem with Turkey, the Turkey's GDP in the first Data Frame shows 1113.561 instead of 1,113,561. Because of that we have 1.11 in the second Data Frame.


### Exercise 3


Load the DataFrame to the CSV file named "Largest_economies.csv"


In [5]:
df.to_csv('./Largest_economies.csv')


---


# Congratulations! You have completed the lab.


Copyright © 2023 IBM Corporation. All rights reserved.
